### PR TITLE
Set Adapter & new specs

### PR DESCRIPTION
**Pull request by VIGER-Valentin**
New Adapter : set_storage.rb
Update of specs for legacy storage & new specs for the new adapter

### CODE REVIEW

File: lib/sidekiq_flow/adapters/set_storage.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+          end
+
+          def workflow_succeeded?(workflow_id)
+            find_workflow(workflow_id).succeeded?
+          end
+
+          def succeed_workflow(workflow_id)
+            workflow_key = find_workflow_key(workflow_id)
+
+            return if already_succeeded?(workflow_key)
```
Comment (franee) [inline lib/sidekiq_flow/adapters/set_storage.rb:93]: we don't need this  call anymore, if the key is already in the set, redis just replies with a 0 / no-op.

File: lib/sidekiq_flow/client.rb [hunk 2]
```diff
@@ -95,7 +93,9 @@ def connection_pool
       private
 
       def adapters
-        @adapters ||= [SidekiqFlow::Adapters::LegacyStorage]
+        #@adapters ||= [SidekiqFlow::Adapters::SetStorage,SidekiqFlow::Adapters::LegacyStorage]
+        @adapters ||= [SidekiqFlow::Adapters::SetStorage]
```
Comment (franee) [inline lib/sidekiq_flow/client.rb:97]: let's use `[SidekiqFlow::Adapters::SetStorage,SidekiqFlow::Adapters::LegacyStorage]` for now.

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 3]
```diff
@@ -0,0 +1,480 @@
+require 'spec_helper'
+
+RSpec.shared_examples '.start_workflow common' do
+  it 'should store the workflow in redis' do
+    expect {
+      subject
+    }.to change { $redis.keys.count }.from(0).to(2)
+
+    expect( $redis.keys.first ).to eq('in-progress')
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:9]: in-progress is our set key, 

should check for `$redis.sismember(IN_PROGRESS, 'workflow-key-here')`

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 4]
```diff
@@ -0,0 +1,480 @@
+require 'spec_helper'
+
+RSpec.shared_examples '.start_workflow common' do
+  it 'should store the workflow in redis' do
+    expect {
+      subject
+    }.to change { $redis.keys.count }.from(0).to(2)
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:7]: there is a `redis`  variable that is a reference of `$redis`, don't use the global one.

see: `let(:redis) { $redis }`

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+      it 'should enqueue a job to Sidekiq' do
+        expect {
+          subject
+        }.to change { Sidekiq::Worker.jobs.count }.from(1).to(2)
+      end
+
+      it 'should enqueue TestTask2 job to Sidekiq' do
+        subject
+        expect(Sidekiq::Worker.jobs.last['args'][1]).to eq(task_klass)
+      end
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:106]: add a spec
`it 'should set task status to enqueued'`

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+    subject { described_class.destroy_succeeded_workflows }
+
+    context 'success' do
+      before do
+        described_class.start_workflow(workflow)
+        SidekiqFlow::Worker.drain
+      end
+
+      it 'should ONLY delete succeeded workflows' do
+        described_class.start_workflow(workflow2)
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:350]: Move `described_class.start_workflow(workflow2)` to the before block.

There are too make checks, but is very hard to follow, let's test one thing at a time (extract to multiple specs).

e.g 

```ruby
it 'changes redis key count' do
  expect {
    subject
  }.to change {
    redis.keys.count)
  }.from(4).to(2)
end
```

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+    end
+
+    context 'success' do
+      it 'should enqueue a job to Sidekiq' do
+        expect {
+          subject
+        }.to change { Sidekiq::Worker.jobs.count }.from(1).to(2)
+      end
+
+      it 'should enqueue TestTask2 job to Sidekiq' do
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:426]: add spec to check that Task status was set to enqueued.

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+    subject { described_class.find_workflow_key(workflow.id) }
+
+    before do
+      described_class.start_workflow(workflow)
+    end
+
+    context 'success' do
+      it 'should return the correct key' do
+        result = subject
+        expect(result).to match(/#{SidekiqFlow.configuration.namespace}\.#{workflow.id}/)
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:444]: just eq()

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+    end
+
+    context 'failure scenario' do
+      it 'returns the wrong key' do
+        SidekiqFlow::Worker.drain
+
+        described_class.start_workflow(workflow)
+
+        result = subject
+        expect(result).to match(/#{SidekiqFlow.configuration.namespace}\.#{workflow.id}/)
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:455]: use eq instead of match.

**Review by franee — CHANGES_REQUESTED** [review]

File: lib/sidekiq_flow/adapters/legacy_storage.rb [hunk 10]
```diff
@@ -83,6 +83,12 @@ def find_first(key_pattern)
       
             result
           end
+
+          def already_started?(workflow_id)
```
[the remark this answers is not in the document]
Comment (VIGER-Valentin) [reply lib/sidekiq_flow/adapters/legacy_storage.rb:87]: client.rb needs it in the `start_workflow` method

File: lib/sidekiq_flow/adapters/set_storage.rb [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            "#{configuration.namespace}.#{workflow_id}"
+          end
+      
+          def connection_pool
+            @connection_pool ||= ConnectionPool.new(size: configuration.concurrency, timeout: configuration.timeout) do
+              Redis.new(url: configuration.redis_url)
+            end
+          end
+
+          def already_started?(workflow_id)
```
[the remark this answers is not in the document]
Comment (VIGER-Valentin) [reply lib/sidekiq_flow/adapters/set_storage.rb:69]: client.rb needs it in the `start_workflow` method

File: spec/sidekiq_flow/adapters/set_storage_spec.rb (hunk 4 above)
Comment (VIGER-Valentin) [reply spec/sidekiq_flow/adapters/set_storage_spec.rb:7]: It's in the `RSpec.shared_examples` block, so no `redis` reference yet. Do I create one ?

File: spec/sidekiq_flow/adapters/set_storage_spec.rb (hunk 3 above)
Comment (VIGER-Valentin) [reply spec/sidekiq_flow/adapters/set_storage_spec.rb:9]: There is no workflow key in this block, did `expect( $redis.smembers(IN_PROGRESS).count ).to eq(1)` instead

File: spec/sidekiq_flow/adapters/set_storage_spec.rb (hunk 4 above)
Comment (franee) [reply spec/sidekiq_flow/adapters/set_storage_spec.rb:7]: shared examples inherit variables from the context that included them.

File: spec/sidekiq_flow/adapters/set_storage_spec.rb (hunk 3 above)
Comment (VIGER-Valentin) [reply spec/sidekiq_flow/adapters/set_storage_spec.rb]: Changed. In the end it's : `expect( redis.sismember(IN_PROGRESS, SidekiqFlow::Client.find_workflow_key(workflow.id)) ).to eq(true)`

File: lib/sidekiq_flow/adapters/set_storage.rb [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            connection_pool.with do |redis|
+              redis.sismember(FINISHED, workflow_key)
+            end
+          end
+    
+        end
+      end
+    end
+  end
+
```
Comment (franee) [inline lib/sidekiq_flow/adapters/set_storage.rb:104]: run rubocop, it will format and remove extra lines.

File: lib/sidekiq_flow/client.rb (hunk 2 above)
Comment (franee) [reply lib/sidekiq_flow/client.rb:97]: resolved but not changed?

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+    subject { described_class.destroy_succeeded_workflows }
+
+    context 'success' do
+      before do
+        described_class.start_workflow(workflow)
+        SidekiqFlow::Worker.drain
+        described_class.start_workflow(workflow2)
+      end
+
+      it 'should ONLY delete succeeded workflows' do
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:359]: break into 2 examples

`it should delete succeeded workflows`
`it should NOT delete in progress workflows`

File: spec/sidekiq_flow/adapters/set_storage_spec.rb [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+
+      it 'should change redis key count' do
+        expect {
+          subject
+        }.to change {
+          redis.keys.count
+        }.from(4).to(2)
+      end
+
+      it 'should NOT find deleted workflows' do
```
Comment (franee) [inline spec/sidekiq_flow/adapters/set_storage_spec.rb:377]: break into 2:

`it should raise error when looking for deleted workflows`
`it should NOT raise an error when looking for in-progress workflows`

**Review by franee — CHANGES_REQUESTED** [review]

File: lib/sidekiq_flow/adapters/set_storage.rb [hunk 15]
```diff
@@ -0,0 +1,104 @@
+module SidekiqFlow
+    module Adapters
+      class SetStorage
+        IN_PROGRESS = 'in-progress'
+        FINISHED = 'finished'
+
+        class << self
+      
+          def store_workflow(workflow, initial=false)
+            workflow_key = find_workflow_key(workflow.id)
```
[the remark this answers is not in the document]
Comment (VIGER-Valentin) [reply lib/sidekiq_flow/adapters/set_storage.rb:10]: So, same change of name for the `legacy_storage.rb` (so the client has the same method to call anyway) ?

File: lib/sidekiq_flow/client.rb (hunk 2 above)
Comment (VIGER-Valentin) [reply lib/sidekiq_flow/client.rb:97]: Yeah, mb, forgot te remove it, but right now I need only the new adapter for the specs. If I keep the legacy one, it will mess with the redis key counts in the specs

Comment (franee) [reply lib/sidekiq_flow/client.rb:97]: > it will mess with the redis key counts in the specs

Yeah, I guess we need to replace/refactor those count specs as well to check for actual keys/values.

File: lib/sidekiq_flow/adapters/legacy_storage.rb [hunk 16]
```diff
@@ -1,126 +1,133 @@
...
-              end
-            end
-      
-            result
-          end
-      
-          private
+      def initialize(configuration)
+        @configuration = configuration
+        @connection_pool ||= ConnectionPool.new(size: configuration.concurrency, timeout: configuration.timeout) do
```
Comment (franee) [inline lib/sidekiq_flow/adapters/legacy_storage.rb:10]: Just move back the connection pool to its original method definition for now as that change will be on a separate ticket.

File: lib/sidekiq_flow/adapters/set_storage.rb [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+  module Adapters
+    class SetStorage
+      IN_PROGRESS = 'in-progress'.freeze
+      FINISHED = 'finished'.freeze
+
+      attr_reader :configuration, :connection_pool
+
+      def initialize(configuration)
+        @configuration = configuration
+        @connection_pool ||= ConnectionPool.new(size: configuration.concurrency, timeout: configuration.timeout) do
```
Comment (franee) [inline lib/sidekiq_flow/adapters/set_storage.rb:11]: same

File: lib/sidekiq_flow/client.rb [hunk 18]
```diff
@@ -1,39 +1,42 @@
 module SidekiqFlow
   class Client
-
     class << self
+      attr_accessor :adapters
+      attr_reader :configuration
```
Comment (franee) [inline lib/sidekiq_flow/client.rb:5]: we already have the method configuration, no need for a reader.

attr_reader is just a shortcut for
```
  def configuration
     @configuration
  end
```

File: lib/sidekiq_flow/client.rb [hunk 19]
```diff
@@ -1,39 +1,42 @@
 module SidekiqFlow
   class Client
-
     class << self
+      attr_accessor :adapters
```
Comment (franee) [inline lib/sidekiq_flow/client.rb:4]: This would probably just be attr_writer (setter) as we already have the private method adapters (getter).

File: spec/sidekiq_flow/client_spec.rb [hunk 20]
```diff
@@ -102,6 +95,11 @@
         subject
         expect(Sidekiq::Worker.jobs.last['args'][1]).to eq(task_klass)
       end
+
```
Comment (franee) [inline spec/sidekiq_flow/client_spec.rb:98]: For the client_spec, please add specs to all public methods.

e.g for store workflow:
```
describe '.store_workflow' do
  it 'should work correctly' do
    subject
    expect(expect(SidekiqFlow::Client.find_workflow('some-id')).to eq(something)
  end
end
```

**Review by franee — CHANGES_REQUESTED** [review]

**Review by franee — APPROVED** [review]

**Review by franee — APPROVED** [review]
